File: shared/android/shared/drive/src/main/kotlin/app/releaf/mobile/data/drive/DriveClientPath.kt
/*
 * DriveClientPath.kt
 *
 * Path-aware extension functions over [DriveClient]. The base interface
 * deals in parent-id + filename pairs; the v2 sync worker wants to say
 * "upload to notepad_entries/2026/04/abc123.json under the Releaf root"
 * and have intermediate folders created as needed.
 *
 * Kept as extensions rather than new interface methods so real and stub
 * DriveClient implementations don't have to re-implement path walking —
 * the semantics are defined once here and compose on top of
 * ensureRootFolder / ensureFolder / uploadJson / downloadBytes /
 * listChildren / trash.
 */

package app.releaf.mobile.data.drive

/**
 * Ensure every folder in [relativePath] exists under [rootFolderId],
 * creating any missing intermediates. Returns the Drive file id of the
 * leaf folder. `relativePath` uses `/` as separator.
 *
 *   ensurePath("notepad_entries/2026/04", rootId) → leaf folder id
 */
suspend fun DriveClient.ensurePath(
    relativePath: String,
    rootFolderId: String,
    accessToken: String,
): String {
    val segments = relativePath.split('/').filter { it.isNotEmpty() }
    var currentId = rootFolderId
    for (seg in segments) {
        currentId = ensureFolder(seg, currentId, accessToken).id
    }
    return currentId
}

/**
 * Upload a JSON payload to [relativePath] under [rootFolderId], creating
 * any missing intermediate folders. Replaces an existing file with the
 * same path.
 */
suspend fun DriveClient.uploadJsonAtPath(
    data: ByteArray,
    relativePath: String,
    rootFolderId: String,
    accessToken: String,
): DriveFile {
    val lastSlash = relativePath.lastIndexOf('/')
    val folderPath = if (lastSlash >= 0) relativePath.substring(0, lastSlash) else ""
    val filename   = if (lastSlash >= 0) relativePath.substring(lastSlash + 1) else relativePath
    val folderId = if (folderPath.isEmpty())
        rootFolderId
    else
        ensurePath(folderPath, rootFolderId, accessToken)
    return uploadJson(data, filename, folderId, accessToken)
}

/**
 * Upload binary bytes (PDF, JPEG, etc.) to [relativePath] under
 * [rootFolderId], creating missing intermediate folders. Replaces an
 * existing file with the same path.
 */
suspend fun DriveClient.uploadBinaryAtPath(
    data: ByteArray,
    contentType: String,
    relativePath: String,
    rootFolderId: String,
    accessToken: String,
): DriveFile {
    val lastSlash = relativePath.lastIndexOf('/')
    val folderPath = if (lastSlash >= 0) relativePath.substring(0, lastSlash) else ""
    val filename   = if (lastSlash >= 0) relativePath.substring(lastSlash + 1) else relativePath
    val folderId = if (folderPath.isEmpty())
        rootFolderId
    else
        ensurePath(folderPath, rootFolderId, accessToken)
    return uploadBinary(data, filename, contentType, folderId, accessToken)
}

/**
 * Download the bytes at [relativePath]. Returns null when any folder
 * along the path is missing, or when the leaf file doesn't exist —
 * matches the "not found" contract the sync worker expects.
 */
suspend fun DriveClient.downloadBytesAtPath(
    relativePath: String,
    rootFolderId: String,
    accessToken: String,
): ByteArray? {
    val segments = relativePath.split('/').filter { it.isNotEmpty() }
    if (segments.isEmpty()) return null

    var currentId = rootFolderId
    for (i in 0 until segments.size - 1) {
        val name = segments[i]
        val folder = listChildren(currentId, accessToken)
            .firstOrNull { it.name == name && it.isFolder }
            ?: return null
        currentId = folder.id
    }
    val filename = segments.last()
    val file = listChildren(currentId, accessToken)
        .firstOrNull { it.name == filename && !it.isFolder }
        ?: return null
    return try {
        downloadBytes(file.id, accessToken)
    } catch (_: DriveError.NotFound) {
        null
    }
}

/**
 * Trash the file at [relativePath]. Returns `true` when the file was
 * found and trashed, `false` otherwise. No-op (returns false) when any
 * ancestor folder is missing.
 */
suspend fun DriveClient.trashAtPath(
    relativePath: String,
    rootFolderId: String,
    accessToken: String,
): Boolean {
    val segments = relativePath.split('/').filter { it.isNotEmpty() }
    if (segments.isEmpty()) return false

    var currentId = rootFolderId
    for (i in 0 until segments.size - 1) {
        val name = segments[i]
        val folder = listChildren(currentId, accessToken)
            .firstOrNull { it.name == name && it.isFolder }
            ?: return false
        currentId = folder.id
    }
    val filename = segments.last()
    val file = listChildren(currentId, accessToken)
        .firstOrNull { it.name == filename && !it.isFolder }
        ?: return false
    return try {
        trash(file.id, accessToken)
        true
    } catch (_: DriveError) {
        false
    }
}

File: shared/ios/ReleafCore/Sources/ReleafCoreDrive/DriveClientPath.swift
/*
 * DriveClientPath.swift
 *
 * Path-aware helpers over `DriveClient`. Mirror of Android's
 * `DriveClientPath.kt` — extension functions that compose on top of
 * the base protocol (ensureRootFolder / ensureFolder / uploadJSON /
 * downloadBytes / listChildren / trash) so neither the real REST
 * client nor the in-memory stub needs extra methods.
 */

import Foundation

public extension DriveClient {

    /// Ensure every folder in [relativePath] exists under [rootFolderId],
    /// creating any missing intermediates. Returns the Drive file id of
    /// the leaf folder.
    func ensurePath(
        _ relativePath: String,
        rootFolderId: String,
        accessToken: String
    ) async throws -> String {
        let segments = relativePath.split(separator: "/").map(String.init)
        var currentId = rootFolderId
        for seg in segments {
            let f = try await ensureFolder(named: seg, parentId: currentId, accessToken: accessToken)
            currentId = f.id
        }
        return currentId
    }

    /// Upload a JSON payload to [relativePath] under [rootFolderId],
    /// creating missing intermediate folders. Replaces an existing file
    /// with the same path.
    @discardableResult
    func uploadJSONAtPath(
        _ data: Data,
        relativePath: String,
        rootFolderId: String,
        accessToken: String
    ) async throws -> DriveFile {
        let segments = relativePath.split(separator: "/").map(String.init)
        guard !segments.isEmpty else {
            throw DriveError.underlying("uploadJSONAtPath: empty path")
        }
        let filename = segments.last!
        let folderSegments = segments.dropLast()
        let folderId: String
        if folderSegments.isEmpty {
            folderId = rootFolderId
        } else {
            folderId = try await ensurePath(
                folderSegments.joined(separator: "/"),
                rootFolderId: rootFolderId,
                accessToken: accessToken
            )
        }
        return try await uploadJSON(data, filename: filename, parentId: folderId, accessToken: accessToken)
    }

    /// Upload binary bytes (PDF, JPEG, etc.) to [relativePath] under
    /// [rootFolderId], creating missing intermediate folders. Replaces
    /// an existing file with the same path.
    @discardableResult
    func uploadBinaryAtPath(
        _ data: Data,
        contentType: String,
        relativePath: String,
        rootFolderId: String,
        accessToken: String
    ) async throws -> DriveFile {
        let segments = relativePath.split(separator: "/").map(String.init)
        guard !segments.isEmpty else {
            throw DriveError.underlying("uploadBinaryAtPath: empty path")
        }
        let filename = segments.last!
        let folderSegments = segments.dropLast()
        let folderId: String
        if folderSegments.isEmpty {
            folderId = rootFolderId
        } else {
            folderId = try await ensurePath(
                folderSegments.joined(separator: "/"),
                rootFolderId: rootFolderId,
                accessToken: accessToken
            )
        }
        return try await uploadBinary(
            data,
            filename: filename,
            contentType: contentType,
            parentId: folderId,
            accessToken: accessToken
        )
    }

    /// Download the bytes at [relativePath]. Returns nil when any
    /// folder along the path is missing or when the leaf file doesn't
    /// exist.
    func downloadBytesAtPath(
        _ relativePath: String,
        rootFolderId: String,
        accessToken: String
    ) async throws -> Data? {
        let segments = relativePath.split(separator: "/").map(String.init)
        guard !segments.isEmpty else { return nil }

        var currentId = rootFolderId
        for i in 0..<(segments.count - 1) {
            let name = segments[i]
            let children = try await listChildren(of: currentId, accessToken: accessToken)
            guard let folder = children.first(where: { $0.name == name && $0.isFolder }) else {
                return nil
            }
            currentId = folder.id
        }
        let filename = segments.last!
        let children = try await listChildren(of: currentId, accessToken: accessToken)
        guard let file = children.first(where: { $0.name == filename && !$0.isFolder }) else {
            return nil
        }
        do {
            return try await downloadBytes(fileId: file.id, accessToken: accessToken)
        } catch DriveError.notFound {
            return nil
        }
    }

    /// Trash the file at [relativePath]. Returns true when found and
    /// trashed, false otherwise.
    @discardableResult
    func trashAtPath(
        _ relativePath: String,
        rootFolderId: String,
        accessToken: String
    ) async throws -> Bool {
        let segments = relativePath.split(separator: "/").map(String.init)
        guard !segments.isEmpty else { return false }

        var currentId = rootFolderId
        for i in 0..<(segments.count - 1) {
            let name = segments[i]
            let children = try await listChildren(of: currentId, accessToken: accessToken)
            guard let folder = children.first(where: { $0.name == name && $0.isFolder }) else {
                return false
            }
            currentId = folder.id
        }
        let filename = segments.last!
        let children = try await listChildren(of: currentId, accessToken: accessToken)
        guard let file = children.first(where: { $0.name == filename && !$0.isFolder }) else {
            return false
        }
        do {
            try await trash(fileId: file.id, accessToken: accessToken)
            return true
        } catch {
            return false
        }
    }
}
